File: code/python/numSquares.py
class Solution:
    def numSquares(self, n: int) -> int:
        if n <= 2:
            return n
        
        lst = []
        i = 1
        while i*i <=n:
            lst.append(i*i)
            i = i+1
            
        count = 0
        nums = set([n])
        while nums:
            count = count + 1
            temp = set()
            for x in nums:
                for y in lst:
                    if x == y:
                        return count
                    if x < y :
                        break
                    temp.add(x-y)
            nums = temp
            
            
        return count

File: code/java/numSquares.java
class Solution {
    public int numSquares(int n) {
        int min = n + 10;
        int[] dp = new int[n+1];
        dp[0] = 0;
        for(int i=1;i<n+1;i++)
            dp[i] = min;
        for(int i=1;i<n+1;i++){
            int j = 1;
            min = n + 10;
            while(j*j<=i){
                min = Math.min(min,dp[i-j*j]+1);
                j = j + 1;
            }
            dp[i] = min;
        }
        return dp[n];
    }
}
